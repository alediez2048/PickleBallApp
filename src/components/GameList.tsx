import React, { useCallback } from "react";
import { View, Text, StyleSheet } from "react-native";
import { FlashList } from "@shopify/flash-list";
import { useGames } from "@/hooks/useGames";
import { GameCard } from "./common/GameCard";
import { prefetch } from "@/utils/prefetch";
import type { Game, SkillLevel } from "@/types/games";

interface GameListProps {
  filters?: {
    skillLevel?: SkillLevel;
    date?: string;
    location?: string;
  };
  onGamePress?: (game: Game) => void;
}

export const GameList = React.memo(function GameList({
  filters,
  onGamePress,
}: GameListProps) {
  const { games, error, isRefreshing, refreshGames } = useGames(filters);

  const handleGamePress = useCallback(
    (game: Game) => {
      // Prefetch game details and booking flow data when a game is pressed
      prefetch.prefetchBookingFlow(game.id);
      onGamePress?.(game);
    },
    [onGamePress]
  );

  const renderItem = useCallback(
    ({ item }: { item: Game }) => (
      <GameCard game={item} onPress={() => handleGamePress(item)} />
    ),
    [handleGamePress]
  );

  if (error) {
    return (
      <View style={styles.errorContainer}>
        <Text style={styles.errorText}>
          {error.message || "An error occurred while loading games"}
        </Text>
      </View>
    );
  }

  return (
    <FlashList
      data={games}
      renderItem={renderItem}
      estimatedItemSize={200}
      onRefresh={refreshGames}
      refreshing={isRefreshing}
      keyExtractor={(item) => item.id}
      contentContainerStyle={styles.container}
    />
  );
});

const styles = StyleSheet.create({
  container: {
    padding: 16,
  },
  errorContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    padding: 16,
  },
  errorText: {
    color: "red",
    textAlign: "center",
  },
});
