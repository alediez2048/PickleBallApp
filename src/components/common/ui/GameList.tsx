import React, { useCallback, useMemo, useEffect } from 'react';
import { View, StyleSheet, Platform } from 'react-native';
import { FlashList, ListRenderItem } from '@shopify/flash-list';
import { Image } from 'expo-image';
import { Game } from '@/types/game';
import { withMemo } from '@/components/hoc/withMemo';
import { useGames } from '@/contexts/GameContext';
import { LoadingSpinner } from './LoadingSpinner';
import { ThemedText } from '@/components/ThemedText';

const blurhash = '|rF?hV%2WCj[ayj[a|j[az_NaeWBj@ayfRayfQfQM{M|azj[azf6fQfQfQIpWXofj[ayj[j[fQayWCoeoeaya}j[ayfQa{oLj?j[WVj[ayayj[fQoff7azayj[ayj[j[ayofayayayj[fQj[ayayj[ayfjj[j[ayjuayj[';

interface GameListProps {
  data?: Game[];
  loading?: boolean;
  error?: string | null;
  onRefresh?: () => void;
  onEndReached?: () => void;
  onGamePress?: (game: Game) => void;
  ListEmptyComponent?: React.ComponentType<any> | React.ReactElement | null;
  ListHeaderComponent?: React.ComponentType<any> | React.ReactElement | null;
  estimatedItemSize?: number;
  prefetchCount?: number;
}

const GameListComponent: React.FC<GameListProps> = ({
  data,
  loading = false,
  error = null,
  onRefresh,
  onEndReached,
  onGamePress,
  ListEmptyComponent,
  ListHeaderComponent,
  estimatedItemSize = 100,
  prefetchCount = 5,
}) => {
  const { games: contextGames, loading: contextLoading, error: contextError, prefetchGame } = useGames();
  const games = data || contextGames;
  const isLoading = loading || contextLoading;
  const hasError = error || contextError;

  // Prefetch next batch of games
  useEffect(() => {
    if (prefetchGame && games.length > 0) {
      const startIndex = Math.max(0, games.length - prefetchCount);
      games.slice(startIndex).forEach(game => {
        prefetchGame(game.id);
      });
    }
  }, [games, prefetchGame, prefetchCount]);

  const keyExtractor = useCallback((item: Game) => item.id, []);

  const renderItem: ListRenderItem<Game> = useCallback(
    ({ item }) => (
      <GameListItem
        game={item}
        onPress={() => onGamePress?.(item)}
      />
    ),
    [onGamePress]
  );

  const EmptyComponent = useMemo(() => {
    if (isLoading) return <LoadingSpinner />;
    if (hasError) {
      return (
        <View style={styles.centerContent}>
          <ThemedText style={styles.errorText}>{hasError}</ThemedText>
        </View>
      );
    }
    if (ListEmptyComponent) return ListEmptyComponent;
    return (
      <View style={styles.centerContent}>
        <ThemedText>No games found</ThemedText>
      </View>
    );
  }, [isLoading, hasError, ListEmptyComponent]);

  return (
    <FlashList
      data={games}
      renderItem={renderItem}
      keyExtractor={keyExtractor}
      estimatedItemSize={estimatedItemSize}
      ListEmptyComponent={EmptyComponent}
      ListHeaderComponent={ListHeaderComponent}
      onRefresh={onRefresh}
      refreshing={isLoading}
      onEndReached={onEndReached}
      onEndReachedThreshold={0.5}
      drawDistance={estimatedItemSize * 10}
      overrideItemLayout={(layout, item) => {
        layout.size = estimatedItemSize;
      }}
      removeClippedSubviews={Platform.OS !== 'web'}
    />
  );
};

const GameListItem = withMemo<{ game: Game; onPress?: () => void }>(
  ({ game, onPress }) => {
    return (
      <View style={styles.itemContainer}>
        {game.location.imageUrl && (
          <Image
            source={game.location.imageUrl}
            style={styles.image}
            placeholder={blurhash}
            contentFit="cover"
            transition={200}
            cachePolicy="memory-disk"
          />
        )}
        <View style={styles.contentContainer}>
          <ThemedText style={styles.title}>{game.title}</ThemedText>
          <ThemedText style={styles.details}>
            {`${game.currentPlayers}/${game.maxPlayers} players • ${game.skillLevel}`}
          </ThemedText>
          <ThemedText style={styles.location}>
            {`${game.location.name} • ${game.location.city}, ${game.location.state}`}
          </ThemedText>
        </View>
      </View>
    );
  },
  (prev, next) => prev.game.id === next.game.id
);

const styles = StyleSheet.create({
  centerContent: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    padding: 20,
  },
  errorText: {
    color: '#DC2626',
    textAlign: 'center',
  },
  itemContainer: {
    flexDirection: 'row',
    padding: 16,
    borderBottomWidth: 1,
    borderBottomColor: '#E5E7EB',
  },
  image: {
    width: 80,
    height: 80,
    borderRadius: 8,
    marginRight: 12,
  },
  contentContainer: {
    flex: 1,
    justifyContent: 'center',
  },
  title: {
    fontSize: 16,
    fontWeight: '500',
    marginBottom: 4,
  },
  details: {
    fontSize: 14,
    color: '#6B7280',
    marginBottom: 2,
  },
  location: {
    fontSize: 14,
    color: '#6B7280',
  },
});

export const GameList = withMemo(GameListComponent); 