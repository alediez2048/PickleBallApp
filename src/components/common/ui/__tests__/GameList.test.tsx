import React from 'react';
import { ReactNode } from 'react';
import renderer from 'react-test-renderer';
import { GameList } from '../GameList';
import type { Game, Location } from '@/types/game';

// Mock the required components
jest.mock('react-native', () => {
  const RN = jest.requireActual('react-native');
  return {
    ...RN,
    View: 'View',
  };
});

// Mock expo-image
jest.mock('expo-image', () => 'Image');

// Create a mock structure factory to replace JSX
const createMockElement = (type: string, props: Record<string, any> = {}, key?: string): any => {
  return {
    type,
    props,
    key
  };
};

// Mock FlashList with a fixed implementation that avoids using JSX
jest.mock('@shopify/flash-list', () => {
  return {
    FlashList: (props: any) => {
      // We're returning a structure instead of JSX
      if (props.data && props.data.length > 0) {
        return createMockElement('View', {
          testID: props.testID || 'flash-list',
          children: [
            props.ListHeaderComponent && createMockElement('View', { 
              children: props.ListHeaderComponent 
            }),
            ...(props.data.map((item: any, index: number) => 
              createMockElement('View', {
                testID: 'list-item-container',
                children: props.renderItem({ item, index })
              }, item.id || index.toString())
            )),
            props.onEndReached && createMockElement('View', {
              testID: 'end-reached',
              onTouchEnd: props.onEndReached
            }),
            props.onRefresh && createMockElement('View', {
              testID: 'refresh-control',
              onTouchEnd: props.onRefresh
            })
          ].filter(Boolean)
        });
      }
      
      return createMockElement('View', {
        testID: props.testID || 'flash-list-empty',
        children: [
          props.ListEmptyComponent,
          props.onRefresh && createMockElement('View', {
            testID: 'refresh-control',
            onTouchEnd: props.onRefresh
          })
        ].filter(Boolean)
      });
    }
  };
});

// Mock the contexts
jest.mock('@/contexts/GameContext', () => ({
  useGames: () => ({
    games: [],
    loading: false,
    error: null,
    prefetchGame: jest.fn(),
  }),
  GameProvider: ({ children }: { children: ReactNode }) => children,
}));

jest.mock('@/contexts/UIContext', () => ({
  UIProvider: ({ children }: { children: ReactNode }) => children,
}));

// Mock ThemedText
jest.mock('@/components/ThemedText', () => ({
  ThemedText: (props: any) => createMockElement('View', {
    testID: 'themed-text',
    style: props.style,
    ...props,
    children: props.children
  }),
}));

// Mock LoadingSpinner
jest.mock('../LoadingSpinner', () => ({
  LoadingSpinner: (props: { message?: string }) => createMockElement('View', {
    testID: 'loading-spinner',
    children: props.message ? createMockElement('View', {
      testID: 'loading-message',
      children: props.message
    }) : null
  }),
}));

const mockLocation: Location = {
  id: '1',
  name: 'Test Location',
  address: '123 Test St',
  city: 'Test City',
  state: 'TS',
  zipCode: '12345',
  coordinates: {
    latitude: 37.7749,
    longitude: -122.4194,
  },
};

const createMockGame = (id: string): Game => {
  // Create a fixed date for testing
  const startDate = new Date('2023-01-01T10:00:00.000Z');
  const endDate = new Date('2023-01-01T12:00:00.000Z');
  
  return {
    id,
    title: `Game ${id}`,
    description: `Game ${id} description`,
    startTime: startDate.toISOString(),
    endTime: endDate.toISOString(),
    location: mockLocation,
    host: {
      id: '1',
      name: 'Host 1',
      email: 'host1@example.com',
      skillLevel: 'Intermediate' as const,
    },
    players: [],
    registeredCount: 4,
    maxPlayers: 8,
    skillLevel: 'Intermediate' as const,
    price: 10,
    imageUrl: 'https://example.com/image1.jpg',
    status: 'Upcoming' as const,
    createdAt: new Date('2023-01-01T09:00:00.000Z').toISOString(),
    updatedAt: new Date('2023-01-01T09:00:00.000Z').toISOString(),
  };
};

describe('GameList', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    
    // Create a consistent date for snapshots
    const mockDate = new Date('2023-01-01T12:00:00.000Z');
    jest.spyOn(global, 'Date').mockImplementation(() => mockDate as any);
  });
  
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('renders correctly with empty list', () => {
    const emptyComponent = <div data-testid="empty-list" />;
    
    const tree = renderer
      .create(
        <GameList
          data={[]}
          ListEmptyComponent={emptyComponent}
        />
      )
      .toJSON();
    
    expect(tree).toMatchSnapshot('empty list');
  });

  it('renders list of games', () => {
    const games = [
      createMockGame('1'),
      createMockGame('2'),
    ];

    const tree = renderer
      .create(<GameList data={games} />)
      .toJSON();
    
    expect(tree).toMatchSnapshot('game list with items');
  });

  it('shows loading state', () => {
    const tree = renderer
      .create(<GameList loading={true} />)
      .toJSON();
    
    expect(tree).toMatchSnapshot('loading state');
  });

  it('shows error state', () => {
    const errorMessage = 'Failed to load games';
    
    const tree = renderer
      .create(<GameList error={errorMessage} />)
      .toJSON();
    
    expect(tree).toMatchSnapshot('error state');
  });

  it('renders with header component', () => {
    const headerComponent = <div data-testid="header-component" />;
    
    const tree = renderer
      .create(
        <GameList 
          data={[createMockGame('1')]} 
          ListHeaderComponent={headerComponent}
        />
      )
      .toJSON();
    
    expect(tree).toMatchSnapshot('with header component');
  });

  it('renders with refresh and end reached handlers', () => {
    const onRefresh = jest.fn();
    const onEndReached = jest.fn();
    
    const tree = renderer
      .create(
        <GameList 
          data={[createMockGame('1')]} 
          onRefresh={onRefresh}
          onEndReached={onEndReached}
        />
      )
      .toJSON();
    
    expect(tree).toMatchSnapshot('with refresh and end reached handlers');
  });
});

export const SkillLevel = {
  Beginner: 'Beginner',
  Intermediate: 'Intermediate',
  Advanced: 'Advanced',
  AllLevels: 'All Levels'
} as const;

export type AppRoutes = {
  '/(tabs)/': undefined;
  '/(tabs)/home': undefined;
  '/(tabs)/games': { category?: string };
  '/games/[id]': { id: string };
  '/games/create': undefined;
  '/activity/[id]': { id: string };
  '/(auth)/login': undefined;
  '/(auth)/register': undefined;
  '/(auth)/forgot-password': undefined;
}; 