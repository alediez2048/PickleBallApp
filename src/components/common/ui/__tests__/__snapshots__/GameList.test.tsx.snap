// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameList renders correctly with empty list: empty list 1`] = `null`;

exports[`GameList renders list of games: game list with items 1`] = `null`;

exports[`GameList renders with header component: with header component 1`] = `null`;

exports[`GameList renders with refresh and end reached handlers: with refresh and end reached handlers 1`] = `null`;

exports[`GameList shows error state: error state 1`] = `null`;

exports[`GameList shows loading state: loading state 1`] = `null`;
